### PR TITLE
Changes to support config.yaml file

### LINKED ISSUE #574: Change slips.conf format to yaml.
Created by Alya Gomaa via monday.com integration. 🎉

### PR DESCRIPTION
**Pull request by lpp5139**
Closes #574 

## Changes proposed

Use slips.yaml file for configuration.

## Steps you followed to test the changes purposed in this PR:

* Add slips.yaml in config folder. 
* Modify all the files that used slips.conf file to support slips.yaml.
* Key files modified are metadata_manager.py, test_p2p.py, arg_parser.py, config_parser.py, test_database.py, docker/P2P-image/DockerFile and docker/macosm1-P2P-image/Dockerfile
* Kept the original configuration file and commented all previous lines, in case of reverting to previous configuration

## Check List (Check all the applicable boxes) 

- [X] My code follows the code style of this project.
- [X] My change requires changes to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [X] This PR does not contain plagiarized content.
- [X] The title of my pull request is a short description of the requested changes.

## Note to reviewers

Few tests fails. 2 main reasons of failure
* modules/threat_intelligence/local_data_files folder is empty, most likely a build error.
*  Bad descriptor error in tests/update_manager.py and hash not matching. 

image

image

### CODE REVIEW

File: docker/P2P-image/Dockerfile [hunk 1]
```diff
@@ -63,7 +63,8 @@ RUN grep -v 'black' install/requirements.txt | xargs pip3 install  --no-cache-di
     && npm install
 
 # set use_p2p to yes in the config file
-RUN sed -i 's/use_p2p = no/use_p2p = yes/g' config/slips.conf
+# RUN sed -i 's/use_p2p = no/use_p2p = yes/g' config/slips.conf
+RUN sed -i 's/use_p2p = no/use_p2p = yes/g' config/slips.yaml
```
Comment (AlyaGomaa) [inline docker/P2P-image/Dockerfile:67]: the purpose of this line in both dockerfiles was to set `use_p2p` to yes instead of no in the config file, now the format of the config file has changes, so we should change the sed cmd as well to use the new format.

File: modules/cesnet/cesnet.py [hunk 2]
```diff
@@ -25,7 +25,7 @@ def init(self):
         self.stop_module = False
 
     def read_configuration(self):
-        """Read importing/exporting preferences from slips.conf"""
+        """Read importing/exporting preferences from slips.conf/slips.yaml"""
```
Comment (AlyaGomaa) [inline modules/cesnet/cesnet.py:28]: we can completely delete slips.conf, once this PR is merged we wont be using it anymore. in case we do neeed it , we cat always get the code back using git, right?

File: slips_files/common/parsers/arg_parser.py [hunk 3]
```diff
@@ -108,7 +110,8 @@ def get_configfile(self):
 
     def parse_arguments(self):
         # Parse the parameters
-        slips_conf_path = os.path.join(os.getcwd(), "config/slips.conf")
+        # slips_conf_path = os.path.join(os.getcwd(), "config/slips.conf")
```
Comment (AlyaGomaa) [inline slips_files/common/parsers/arg_parser.py:113]: hey feel free to delete code that is no longer needed instead of commenting it :)) we can always find it again if necessary

File: config/slips.yaml [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+
+# enable memory profiling [yes,no]
+   memory_profiler_enable : no
+
+# set profiling mode [dev,live]
+   memory_profiler_mode : live
+
+# profile all subprocesses [yes,no]
+   memory_profiler_multiprocess : yes
+
```
Comment (AlyaGomaa) [inline config/slips.yaml:436]: in the latest develop, there's a parameter here for the web interface port, slips willl complain when it's missing

**Review by AlyaGomaa — COMMENTED** [review]
Hey @lpp5139 thank you so much for all the hard work! we really appreciate it.

I pulled the latest develop into your branch, and left you some comments, once they're resolved the tests will pass and we'lll be able to merge your PR easily.

Keep up the good work!!

**Comment by lpp5139** [comment]
**Changed in latest commit** 
* Modified sed command in DockerFile
* Added interface port in slips.yaml file. Commented out slips.conf file. 
* Removed slips.conf lines from some file
* Couldn't test it on P2P image as P2p4slips folder is empty, but it passed the sed command without any error.

**Comment by AlyaGomaa** [comment]
Hello @lpp5139 the failing unit tests were probably because of the default value of threat intelligence local files. i fixed it here https://github.com/stratosphereips/StratosphereLinuxIPS/pull/587/commits/7aa77065f719261f66f6675df431141b57c7aeb1
now one more step to make the tests pass and its the following:
in slips, we have 2 inetgration tests for testing the config file. 
these tests are the ones failing in `CI-staging / config_files_integration_tests`, you can find them here `tests/integration_tests/test_config_files.py`
and they're using the 2 config files in the same dir, once these files are updated to use the latest yaml format, i think all tests will pass no issue
do you wanna update them? 
I will test the dockerfiles and get back to you

File: slips_files/common/parsers/config_parser.py [hunk 5]
```diff
@@ -27,14 +30,14 @@ def __init__(self):
...
-        reads slips configuration file, slips.conf is the default file
+        reads slips configuration file, slips.conf/slips.yaml is the default file
         """
         config = configparser.ConfigParser(
             interpolation=None, comment_prefixes="#"
         )
         try:
             with open(self.configfile) as source:
-                config.read_file(source)
+                yaml.safe_load(source)
```
Comment (AlyaGomaa) [inline slips_files/common/parsers/config_parser.py:40]: now after parsing the file, the return of this line is never used, therefore the config.yaml is never used by slips

File: slips_files/common/parsers/config_parser.py [hunk 6]
```diff
@@ -469,7 +472,7 @@ def local_ti_data_path(self):
         return self.read_configuration(
```
Comment (AlyaGomaa) [inline slips_files/common/parsers/config_parser.py:472]: the read_configuration function is not modified by youright? this function should be modified to read from the parsed yaml because now its expecting to read from the old config files

File: slips_files/common/parsers/config_parser.py [hunk 7]
```diff
@@ -82,7 +85,7 @@ def web_interface_port(self) -> int:
 
     def get_entropy_threshold(self):
         """
-        gets the shannon entropy used in detecting C&C over DNS TXT records from slips.conf
+        gets the shannon entropy used in detecting C&C over DNS TXT records from slips.conf/slips.yaml
```
Comment (AlyaGomaa) [inline slips_files/common/parsers/config_parser.py:88]: all comments saying slips.conf/slips.yaml should say slips.yaml only, since slips.conf will be deleted for good. and also you should delete the slips.conf from git in your PR, we will no longer need it right?

**Review by AlyaGomaa — COMMENTED** [review]
Hey @lpp5139 the CI was taking so long because slips is not working. this indicates that you didn't test your changes.
so for example, the first thing i did to debug why CI is failing was just run slips normally like this
`./slips.py -e 1 -f  dataset/test6-malicious.suricata.json `
and it doesn't even start all the modules, because it's not parsing the old slips.conf or the new one
So it's always a good idea to run slips in like 1 or 2 files from the dataset/ directory and thenn, run the tests locally  before submitting a PR, because we cant debug CI in real time right?
thankss

**Comment by lpp5139** [comment]
Hey @AlyaGomaa I modified the file as per your instruction and tested it against "./slips.py -e 1 -f  dataset/test6-malicious.suricata.json" and got the following result(image 1, everything else passed). On testing on other test cases, I am getting a consistent error which is regarding database manger. Can you please help me with this? 

image
image

**Comment by AlyaGomaa** [comment]
hello @lpp5139 only thing that is left is to update the config files integration tests, once this is done CI will pass and i'll be able to merge

**Comment by lpp5139** [comment]
Pushed the code for fixing integration test. I got the error related to database, which is consistent my environment. No error with configuration files. 
image

**Comment by AlyaGomaa** [comment]
this is related to the new config files that you added,  it's not related to the db or slips code, is it possible that you debug this further?

**Comment by lpp5139** [comment]
Hey @AlyaGomaa, I updated the test file and test2.yaml passes the test case. Still struggling to pass the first one. Can you please help me with that? 
My docker only supports port 6379, so I need to test with that. As we are reading pcap files, the error could be that I am testing it on wrong port. 

image

**Comment by AlyaGomaa** [comment]
hi @lpp5139 when profiles is == 0 in any failing test, this means that slips didn't run in this test, some error occurred that made slips not run and the analysis finished with 0 profiles
here's how i would debug this
1. check the output directory of the failing test, it is test_configuration_file/ in the screenshot above for any errors
2. run the command using in this test by hand, with the same config file used and everything, you will get the error printed to your cli and you can check it from there

my guess is that slips didn't read the config file in this test correctly
Also docker works with all ports, not just 6379, what do you mean by that?? anyways the test that is failing seems to be using port 6379

File: tests/integration_tests/test_config_files.py [hunk 8]
```diff
@@ -1,136 +1,136 @@
...
+    output_dir = create_output_dir(output_dir)
+    output_dir = create_output_dir(output_dir)
+
+    output_file = os.path.join(output_dir, "slips_output.txt")
+    command = (
+        f"./slips.py "
+        f"-t  -e 1 "
+        f"-f {pcap_path} "
+        f"-o {output_dir} "
+        f"-c /StratosphereLinuxIPS/tests/integration_tests/test.yaml "
```
Comment (AlyaGomaa) [inline tests/integration_tests/test_config_files.py:56]: why did you add /StratosphereLinuxIPS/ here? it should be relative, should start with tests/

File: tests/integration_tests/test_config_files.py [hunk 9]
```diff
@@ -1,136 +1,136 @@
...
+    zeek_output_dir = database.get_zeek_output_dir()[2:]
+    assert zeek_output_dir not in os.listdir()
+
+    # test store_a_copy_of_zeek_files
+    assert "zeek_files" in os.listdir(output_dir)
+
+    # test metadata_dir
+    assert "metadata" in os.listdir(output_dir)
+    metadata_path = os.path.join(output_dir, "metadata")
+    for file in ("/StratosphereLinuxIPS/config/test.yaml", "/StratosphereLinuxIPS/config/whitelist.conf", "running_slips_info.txt"):
```
Comment (AlyaGomaa) [inline tests/integration_tests/test_config_files.py:91]: using absolute file paths will cause CI to not be able to find these files, or if another developer has slips code installed in somewhere else other than /StratosphereLinuxIPS/, the tests will fail for them too. so always use relative file paths

**Review by AlyaGomaa — COMMENTED** [review]
i left some comment for you to check @lpp5139

**Comment by AlyaGomaa** [comment]
woooohooo finally merged! @lpp5139 
thank you for your effort!
